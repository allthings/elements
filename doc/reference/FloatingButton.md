<!-- 
This is an auto-generated markdown. 
You can change it in "src/FloatingButton/FloatingButton.tsx" and run build:docs to update this file.
-->
# FloatingButton

## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|color|string|
|type|"submit", "button", "reset"|
|disabled|boolean|
|inProgress|boolean|
|disabledColor|string|
