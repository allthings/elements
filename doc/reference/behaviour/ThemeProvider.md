<!-- 
This is an auto-generated markdown. 
You can change it in "src/behaviour/ThemeProvider.jsx" and run build:docs to update this file.
-->
# ThemeProvider
All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
The ThemeProvider allows you to define the default colors for most elements.

**Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|root|bool|
|theme|shape|<br>Default: {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}
|children|node|
