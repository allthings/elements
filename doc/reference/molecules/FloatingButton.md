<!-- 
This is an auto-generated markdown. 
You can change it in "src/molecules/FloatingButton.jsx" and run build:docs to update this file.
-->
# FloatingButton

## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|color **(required)**|string|
|disabled|bool|
|inProgress|bool|<br>Default: false
|children|node|
|disabledColor **(required)**|string|
