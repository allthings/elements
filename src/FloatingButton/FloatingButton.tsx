import { css } from 'glamor'
import { color as col, lightness } from 'kewler'
import React, { HTMLAttributes } from 'react'
import Absolute from '../Absolute'
import ListSpinner from '../ListSpinner'
import Theme from '../Theme'
import View from '../View'

const buttonStyle = css({
  width: '100%',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100%',
  background: 'transparent',
  border: 0,
  ':active': {
    background: 'rgba(0, 0, 0, 0.15)',
  },
  ':disabled': {
    background: 'rgba(0, 0, 0, 0.15)',
  },
})

interface IFloatingButtonProps extends HTMLAttributes<HTMLButtonElement> {
  color?: string
  type?: 'submit' | 'button' | 'reset'
  disabled?: boolean
  /* True to use the button that something is in progress */
  inProgress?: boolean
  disabledColor?: string
}

const FloatingButton = ({
  color = 'primary',
  disabled,
  disabledColor = 'lightGrey',
  inProgress = false,
  children,
  ...props
}: IFloatingButtonProps) => {
  const isDisabled = disabled || inProgress

  return (
    <Theme>
      {({ colorize }) => (
        <View>
          <View style={{ height: 50 }} />
          <Absolute
            alignH="center"
            alignV="center"
            bottom={0}
            direction="row"
            flex="flex"
            {...css({
              // we only like to change the color if the button is disabled by using the prop
              backgroundColor: colorize(disabled ? disabledColor : color),
              boxShadow: '0px -2px 10px 0px rgba(0, 0, 0, 0.2)',
              cursor: isDisabled ? 'default' : 'pointer',
              height: 50,
              overflow: 'hidden',
              transition: '250ms ease-in-out',
              width: '100%',
              ':hover': {
                background: colorize(
                  disabled
                    ? disabledColor
                    : color.indexOf('#') !== -1
                    ? col(color, lightness(-10))
                    : color,
                ),
              },
            })}
          >
            <button {...buttonStyle} {...props} disabled={isDisabled}>
              {inProgress ? <ListSpinner radius={30} /> : children}
            </button>
          </Absolute>
        </View>
      )}
    </Theme>
  )
}

export default FloatingButton
