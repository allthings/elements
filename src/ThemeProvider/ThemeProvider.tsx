import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme: ITheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  readonly primary: string
  readonly text: string
  readonly secondaryText: string
  readonly titleColor: string
  readonly contrast: string
  readonly warn: string
  readonly disabled: string
  readonly background: string
  readonly textOnBackground: string
  readonly [key: string]: string
}

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = defaultTheme,
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider
export const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer
