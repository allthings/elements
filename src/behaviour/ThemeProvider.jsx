import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast, Subscriber } from 'react-broadcast'
import { ColorPalette } from '@allthings/colors'

export const THEME_CHANNEL = 'theme'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

/**
* All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
* The ThemeProvider allows you to define the default colors for most elements.
*
* **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
**/
export default class ThemeProvider extends React.PureComponent {
  static propTypes = {
    root: PropTypes.bool,
    theme: PropTypes.shape({
      primary: PropTypes.string,
      text: PropTypes.string,
      secondaryText: PropTypes.string,
      titleColor: PropTypes.string,
      contrast: PropTypes.string,
      warn: PropTypes.string,
      disabled: PropTypes.string,
      background: PropTypes.string,
      textOnBackground: PropTypes.string,
    }),
    children: PropTypes.node,
  }

  static defaultProps = {
    theme: defaultTheme,
  }

  static contextTypes = Subscriber.contextTypes

  render() {
    let root = true
    if (this.context.broadcasts && this.context.broadcasts[THEME_CHANNEL]) {
      root = false
    }
    const { theme, children } = this.props
    return root ? (
      <Broadcast
        channel={THEME_CHANNEL}
        value={{
          ...ThemeProvider.defaultProps.theme,
          ...theme,
        }}
      >
        {children}
      </Broadcast>
    ) : (
      <Subscriber channel={THEME_CHANNEL}>
        {previousTheme => (
          <Broadcast
            channel={THEME_CHANNEL}
            value={{
              ...ThemeProvider.defaultProps.theme,
              ...previousTheme,
              ...theme,
            }}
          >
            {children}
          </Broadcast>
        )}
      </Subscriber>
    )
  }
}

export const ThemeShape = ThemeProvider.propTypes.theme

export const withTheme = (mapThemeToProps, displayName) => WrappedComponent =>
  // eslint-disable-next-line
  class extends React.PureComponent {
    static displayName = displayName || WrappedComponent.displayName
    static component = WrappedComponent

    renderComponent = theme => {
      const props = !mapThemeToProps
        ? { theme }
        : mapThemeToProps(theme, this.props)

      return <WrappedComponent {...this.props} {...props} />
    }

    render() {
      return (
        <Subscriber channel={THEME_CHANNEL}>{this.renderComponent}</Subscriber>
      )
    }
  }
