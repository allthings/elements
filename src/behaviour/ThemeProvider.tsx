import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  primary: string
  text: string
  secondaryText: string
  titleColor: string
  contrast: string
  warn: string
  disabled: string
  background: string
  textOnBackground: string
  [key: string]: string
}

const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = {},
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider

type mapThemeToPropsType = (theme: ITheme, props: any) => any

function createThemeConsumer<T>(
  mapThemeToProps: mapThemeToPropsType,
  WrappedComponent: React.ComponentType<T>,
  props: Omit<T, 'theme'>
) {
  return (theme: ITheme) => {
    const extraProps = !mapThemeToProps
      ? { theme }
      : mapThemeToProps(theme, props)

    return <WrappedComponent {...props} {...extraProps} />
  }
}

export function withTheme(mapThemeToProps: mapThemeToPropsType) {
  return function<T extends { theme: ITheme }>(
    WrappedComponent: React.ComponentType<T>
  ): React.ComponentType<Omit<T, 'theme'>> {
    return function(props: Omit<T, 'theme'>) {
      return (
        <ThemeConsumer>
          {createThemeConsumer<T>(mapThemeToProps, WrappedComponent, props)}
        </ThemeConsumer>
      )
    }
  }
}
