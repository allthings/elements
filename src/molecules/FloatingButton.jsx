import React from 'react'
import Absolute from '../atoms/Absolute'
import { withTheme } from '../behaviour/ThemeProvider'
import PropTypes from 'prop-types'
import { css } from 'glamor'
import View from '../atoms/View'

const buttonStyle = css({
  width: '100%',
  display: 'flex',
  alignItem: 'center',
  justifyContent: 'center',
  height: '100%',
  background: 'transparent',
  border: 0,
})

const availableSizes = {
  xs: 30,
  s: 40,
  m: 50,
  l: 60,
  xl: 70,
}

class FloatingButton extends React.Component {
  static propTypes = {
    color: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    disabledColor: PropTypes.string.isRequired,
    size: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl']),
  }

  static defaultProps = {
    size: 'l',
  }

  render() {
    const { color, disabled, disabledColor, size, ...props } = this.props
    return (
      <View>
        <View style={{ height: availableSizes[size] }} />
        <Absolute
          alignH="center"
          alignV="center"
          bottom={0}
          direction="row"
          flex="flex"
          {...css({
            backgroundColor: disabled ? disabledColor : color,
            boxShadow: '0px -2px 10px 0px rgba(0, 0, 0, 0.2)',
            height: availableSizes[size],
            overflow: 'hidden',
            width: '100%',
          })}
        >
          <button {...buttonStyle} {...props} {...css({cursor: disabled ? 'default' : 'pointer'})} />
        </Absolute>
      </View>
    )
  }
}

const mapThemeToProps = theme => ({
  color: theme.primary,
  disabledColor: 'lightGray',
})

export default withTheme(mapThemeToProps)(FloatingButton)
