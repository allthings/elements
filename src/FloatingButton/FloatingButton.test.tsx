import React from 'react'
import renderer from 'react-test-renderer'
import ThemeProvider from '../ThemeProvider'
import FloatingButton from './FloatingButton'

test('Floating button renders without error', () => {
  const tree = renderer
    .create(
      <ThemeProvider>
        <FloatingButton>Hello World</FloatingButton>
      </ThemeProvider>,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
