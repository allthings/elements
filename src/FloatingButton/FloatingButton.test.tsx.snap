// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Floating button renders without error 1`] = `
.css-0,
[data-css-0] {
  width: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  background: transparent;
  border: 0;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
}

.css-0:active,
[data-css-0]:active {
  background: rgba(0, 0, 0, 0.15);
}

.css-0:disabled,
[data-css-0]:disabled {
  background: rgba(0, 0, 0, 0.15);
}

.css-3,
[data-css-3] {
  background-color: #3598db;
  box-shadow: 0px -2px 10px 0px rgba(0, 0, 0, 0.2);
  cursor: pointer;
  height: 50px;
  overflow: hidden;
  transition: 250ms ease-in-out;
  width: 100%;
  -webkit-transition: 250ms ease-in-out;
  -moz-transition: 250ms ease-in-out;
}

.css-3:hover,
[data-css-3]:hover {
  background: #3598db;
}

.css-5,
[data-css-5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-6,
[data-css-6] {
  position: absolute;
  bottom: 0;
}

.css-7,
[data-css-7] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 1;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 1;
}

<div
  data-css-5=""
>
  <div
    data-css-5=""
    style={
      Object {
        "height": 50,
      }
    }
  />
  <div
    data-css-7=""
    data-css-6=""
    data-css-3=""
  >
    <button
      data-css-0=""
      disabled={false}
    >
      Hello World
    </button>
  </div>
</div>
`;
